src/main.ts

```diff
@@ -1,127 +1,192 @@
-				this.scrollUp(leaf);
-				return false;
-			}
-			return true;
-		});
-		app.keymap.pushScope(this.navScope);
-
-		console.log('Vim Reading View Navigation loaded.');
-	}
-
-	displayValue(leaf: MarkdownView): string {
-		return leaf.contentEl
-			.getElementsByClassName('markdown-reading-view')[0]
-			.getElementsByClassName('document-search-container')[0]
-			.getCssPropertyValue('display');
-	}
-
-	async onunload() {
-		app.keymap.popScope(this.navScope);
-		console.log('Vim Reading View Navigation unloaded.');
-	}
-
-	getScroll(leaf: MarkdownView): number {
-		return leaf.previewMode.getScroll();
-	}
-
-	scrollDown(leaf: MarkdownView) {
-		const scroll = this.getScroll(leaf);
-		leaf.previewMode.applyScroll(scroll + this.settings.scrollDifference);
-	}
-
-	scrollUp(leaf: MarkdownView) {
-		const scroll = this.getScroll(leaf);
-		leaf.previewMode.applyScroll(scroll - this.settings.scrollDifference);
-	}
-
-	async loadSettings() {
-		this.settings = Object.assign(
-			{},
-			DEFAULT_SETTINGS,
-			await this.loadData()
-		);
-	}
-
-	async saveSettings() {
-		await this.saveData(this.settings);
-	}
+    settings: VimScrollSetting;
+    navScope: Scope;
+
+    async onload() {
+        await this.loadSettings();
+        this.addSettingTab(new VimScrollSettingTab(this.app, this));
+        this.navScope = new Scope(app.scope);
+
+        // Scroll down
+        this.navScope.register([], 'j', (evt: KeyboardEvent) => {
+            const leaf = app.workspace.getActiveViewOfType(MarkdownView);
+            if (
+                leaf.getMode() === 'preview' &&
+                this.displayValue(leaf) === 'none'
+            ) {
+                this.scrollDown(leaf);
+                return false;
+            }
+            return true;
+        });
+
+        // Scroll up
+        this.navScope.register([], 'k', (evt: KeyboardEvent) => {
+            const leaf = app.workspace.getActiveViewOfType(MarkdownView);
+            if (
+                leaf.getMode() === 'preview' &&
+                this.displayValue(leaf) === 'none'
+            ) {
+                this.scrollUp(leaf);
+                return false;
+            }
+            return true;
+        });
+
+        // Jump to top
+        let keyArray: string[] = [];
+        const jumpTopEvent = (event: KeyboardEvent) => {
+            const leaf = app.workspace.getActiveViewOfType(MarkdownView);
+            if (
+                leaf.getMode() === 'preview' &&
+                this.displayValue(leaf) === 'none'
+            ) {
+                event.preventDefault();
+                if (event.key != 'g') {
+                    keyArray = [];
+                } else {
+                    keyArray.push(event.key);
+                    if (keyArray.length === 2) {
+                        leaf.previewMode.applyScroll(0);
+                        keyArray = [];
+                    }
+                }
+            }
+        }
+        addEventListener('keydown', jumpTopEvent);
```

Comment:
Because it doesn't use Obsidian's hotkey system, this also gets triggered when the settings pane is open. When the current focused pane is in reading view, this has the effect that no text can be entered in text fields in the settings.

Reply:
Yes... I noticed it too (and it might also be the reason for #3, not sure).

Reply:
I suppose the best way to fix this is to register an event `(app.workspace.on('layout-change')` (the other event listeners have some deficiencies when trying to target each pane) and to attach all the event listeners hotkeys to a markdown pane.

Reply:
Ill give it a shot, but it might take a while for me to learn and implement it (my js/ts is limited).

Reply:
I chose the global way because I didn't want to manage it for each pane. 

Although it should be better to better use `active-leaf-change` for it. I was thinkig from another perspective, but these commands are needed only for markdown panes, so that event is better. 

![image](https://user-images.githubusercontent.com/83140328/183970365-3bbd9650-a593-4672-a2ab-69fa927c600b.png)

You can register Events on the view like this. You'll probably need `registerScopeEvent` for the Obsidian event types and `register` for the `addEventListener` (or maybe register the `removeEventListener` function after adding it for the pane, so that Obsidian unloads it when the pane gets destroyed. Although you'll also need to make sure not to add more than one event listener with `active-leaf-change`).

That would probably involve removing and adding it each time `active-leaf-change` fires unless you keep a history of leaf IDs.

![image](https://user-images.githubusercontent.com/83140328/183972964-ba4f9e38-7d99-42aa-8472-3b02008860f0.png)


Considering all of this state management and that Obsidian doesn't provide a hook for when new panes get created, it might be easiest to just go back and use Obsidian's way and just `g` instead of `gg`.

Reply:
Or maybe add an event listener after `g` gets invoked the first time using Obsidian's event system, if not a `g` follows, remove the listener and do nothing.

Reply:
> Or maybe add an event listener after `g` gets invoked the first time using Obsidian's event system, if not a `g` follows, remove the listener and do nothing.

Implemented this, and it appears to work (text can be entered in settings pane too).